File: backend/models/schemas.py
from pydantic import BaseModel
from typing import Optional, List, Dict

#0. LLM Request
class LLMRequest(BaseModel):
    provider: str
    model: str
    api_key: str


#1. Explain Errors
class ExplainErrorsRequest(BaseModel):
    code: str
    error_message: str
    language: Optional[str] = "python"
    llm_request: LLMRequest
    
class ExplainErrorsResponse(BaseModel):
    explanation: str
    possible_causes: Optional[List[str]] = []

#2. Suggest Fixes
class SuggestFixesRequest(BaseModel):
    code: str
    error_message: str
    user_request: str
    language: Optional[str] = "python"
    llm_request: LLMRequest
    
class SuggestFixesResponse(BaseModel):
    fixed_code: str
    fixes: Optional[List[str]] = []
    differences: Optional[List[str]] = []
    
#3. Generate TestCases
class GenerateTestCasesRequest(BaseModel):
    code: str
    code_explanation: Optional[str] = None
    num_testcases: str
    language: Optional[str] = "python"
    llm_request: LLMRequest
    
class TestCase(BaseModel):
    input: Dict[str, str]
    expected_output: str
    explanation: Optional[str] = None
        
class GenerateTestCasesResponse(BaseModel):
    testcases: List[TestCase]
    
#4. Code Walkthrough
class CodeWalkthroughRequest(BaseModel):
    code: str
    focus_on: Optional[str] = None 
    language: Optional[str] = "python"
    llm_request: LLMRequest

class CodeSegmentExplanation(BaseModel):
    segment: str
    step: str

class CodeWalkthroughResponse(BaseModel):
    walkthrough: List[CodeSegmentExplanation]
    

File: src/schemas.ts
// 0. LLMRequest
export interface LLMRequest{
    provider: string,
    model: string,
    api_key: string
}

//1. Explain Errors
export interface ExplainErrorsRequest{
    code: string,
    error_message: string,
    language?: string,
    llm_request: LLMRequest
}

export interface ExplainErrorsResponse{
    explanation: string,
    possible_causes?: string[]
}

//2. Suggest Fixes
export interface SuggestFixesRequest{
    code: string,
    error_message: string,
    user_request: string
    language?: string,
    llm_request: LLMRequest
}

export interface SuggestFixesResponse{
    fixed_code: string,
    fixes?: string[],
    differences?: string[]
}

//3. Generate Testcases
export interface GenerateTestCasesRequest{
    code: string,
    code_explanation?: string,
    num_testcases: string,
    language?: string,
    llm_request: LLMRequest
}

export interface TestCase{
    input: Record<string, any>,
    expected_output: string,
    explanation?: string
}

export interface GenerateTestCasesResponse{
    testcases: TestCase[]
}

//4. Code Walkthrough
export interface CodeWalkthroughRequest{
    code: string,
    focus_on?: string,
    language?: string,
    llm_request: LLMRequest
}

export interface CodeSegmentExplanation{
    segment: string,
    step: string
}

export interface CodeWalkthroughResponse{
    walkthrough: CodeSegmentExplanation[]
}
